source/common/router/router.cc

```diff
@@ -679,84 +748,132 @@ Filter::streamResetReasonToResponseFlag(Http::StreamResetReason reset_reason) {
   NOT_REACHED_GCOVR_EXCL_LINE;
 }
 
-void Filter::handleNon5xxResponseHeaders(const Http::HeaderMap& headers, bool end_stream) {
+void Filter::handleNon5xxResponseHeaders(const Http::HeaderMap& headers,
+                                         UpstreamRequest* upstream_request, bool end_stream) {
   // We need to defer gRPC success until after we have processed grpc-status in
   // the trailers.
   if (grpc_request_) {
     if (end_stream) {
       absl::optional<Grpc::Status::GrpcStatus> grpc_status = Grpc::Common::getGrpcStatus(headers);
       if (grpc_status &&
           !Http::CodeUtility::is5xx(Grpc::Utility::grpcToHttpStatus(grpc_status.value()))) {
-        upstream_request_->upstream_host_->stats().rq_success_.inc();
+        upstream_request->upstream_host_->stats().rq_success_.inc();
       } else {
-        upstream_request_->upstream_host_->stats().rq_error_.inc();
+        upstream_request->upstream_host_->stats().rq_error_.inc();
       }
     } else {
-      upstream_request_->grpc_rq_success_deferred_ = true;
+      upstream_request->grpc_rq_success_deferred_ = true;
     }
   } else {
-    upstream_request_->upstream_host_->stats().rq_success_.inc();
+    upstream_request->upstream_host_->stats().rq_success_.inc();
   }
 }
 
-void Filter::onUpstream100ContinueHeaders(Http::HeaderMapPtr&& headers) {
+void Filter::onUpstream100ContinueHeaders(Http::HeaderMapPtr&& headers,
+                                          UpstreamRequest* upstream_request) {
   ENVOY_STREAM_LOG(debug, "upstream 100 continue", *callbacks_);
 
-  downstream_response_started_ = true;
+  if (!downstream_response_started_) {
+    downstream_response_started_ = true;
+    final_upstream_request_ = upstream_request;
+    resetOtherUpstreams(upstream_request);
+  }
   // Don't send retries after 100-Continue has been sent on. Arguably we could attempt to do a
   // retry, assume the next upstream would also send an 100-Continue and swallow the second one
   // but it's sketchy (as the subsequent upstream might not send a 100-Continue) and not worth
   // the complexity until someone asks for it.
   retry_state_.reset();
 
-  callbacks_->encode100ContinueHeaders(std::move(headers));
+  if (final_upstream_request_ == upstream_request) {
+    callbacks_->encode100ContinueHeaders(std::move(headers));
+  }
+}
+
+void Filter::resetOtherUpstreams(UpstreamRequest* upstream_request) {
+  UpstreamRequest* upstream_request_tmp;
+  for (auto it = upstream_requests_.cbegin(); it != upstream_requests_.cend(); it++) {
+    upstream_request_tmp = it->get();
+    if (upstream_request_tmp != upstream_request) {
+      if (!upstream_request_tmp->encode_complete_ || !upstream_request_tmp->decode_complete_) {
```

Comment:
Is this just implicitly checking if the request is not the winner? Is there a more explicit way? If not, add a comment?

Reply:
bump

Reply:
this was unnecessary because it's checked in resetStream, so removed it